### PR TITLE
Side navigation bar

### PR DESCRIPTION
**Pull request by tcstewar**
This reworks the top toolbar, adding a full side navigation bar.

The file open menu is now done with the sidebar, rather than in its own window (and thus no longer disappears if your mouse cursor goes outside it).  Many of the options in the top toolbar have been moved to the side bar, clearing up the top menu a bit.

Eventually, this will also provide the basis for drag-and-drop support, and we may want to merge the config menu into this nav bar (rather than having it as a separate modal popup).

### CODE REVIEW

**Comment by Seanny123** [comment]
Is this ready for review or does other stuff need to be changed?

**Comment by tcstewar** [comment]
> Is this ready for review or does other stuff need to be changed?

If you'd like to start taking a look at it, that'd be great -- I think the only functional change I want to make is to have the side bar hidden on startup (rather than shown on startup).

File: nengo_gui/static/components/netgraph.js [hunk 1]
```diff
@@ -16,6 +16,8 @@ Nengo.NetGraph = function(parent, args) {
     this.offsetX = 0;          // global x,y pan offset
     this.offsetY = 0;
 
+    this.overide_positions = {};
```
Comment (Seanny123) [inline nengo_gui/static/components/netgraph.js]: A comment explaining what this is used for would be appreciated. From what I can tell it's used for positioning the side-menu, but I'm not clear on why that required a new property to be created.

File: nengo_gui/static/side_menu.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
+
+/**
+ * Menu for the side of the GUI
+ *
+ * SideMenu constructor is written into HTML file by python and called
```
Comment (Seanny123) [inline nengo_gui/static/side_menu.js]: Once this component gets merged, it should be added to the overview documentation. Do we want to log issues for that documentation on GitHub or just in Google Drive? Does anyone know what I'm talking about when I say "overview documentation"?

File: nengo_gui/static/side_menu.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  $('#Toggle_drag_drop')[0].addEventListener('click', function () {
+      self.toggle_display_tab("#Component_menu");
+  });
+
+  //Toggles Side Nav
+  $('.side_toggle').on('click',function(){
+      self.toggle_side_nav();
+  });
+
+    // Handles Accordions
```
Comment (Seanny123) [inline nengo_gui/static/side_menu.js]: nitpick: extra space

File: nengo_gui/static/side_menu.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            counter++;
+        }
+    }
+}
+
+//** Toggle show/hide of a given tab **/
+Nengo.SideMenu.prototype.toggle_display_tab = function(tab_id){
+    var self = this;
+    var index = 0;
+    while(index < self.sub_menus.length){
```
Comment (Seanny123) [inline nengo_gui/static/side_menu.js]: Is this while-loop totally necessary? Could the mapping between tab_id and index be saved in a dictionary?

File: nengo_gui/static/side_menu.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        $(self.sub_menus[index].id).hide();
+        $(self.sub_menus[index].tab).hide();
+    } else {
+        self.sub_menus[index].hidden = false;
+        $(self.sub_menus[index].id).removeClass('hidden');
+        $(self.sub_menus[index].id).show();
+        $(self.sub_menus[index].tab).show();
+    }
+    self.refresh();
+    var shown_tabs = self.sub_menus.filter(function(item){
```
Comment (Seanny123) [inline nengo_gui/static/side_menu.js]: nitpick: is this correct spacing for `function(item) {`

**Comment by Seanny123** [comment]
Sorry to bring up styling again, but are we going to worry about styling at all for this pull request? Like, even for the newly created files? Or if I'm reviewing, should I just focus on logic?

**Comment by tcstewar** [comment]
> Sorry to bring up styling again, but are we going to worry about styling at all for this pull request? Like, even for the newly created files? Or if I'm reviewing, should I just focus on logic?

Good question on the styling.  Logic is definitely more important, but the spacing and styling is also important (yay readability).  I haven't done a pass through to fix to formatting yet, so feel free to note points where you see it.  If there's tons of them, then don't bother catching all of them -- just give a general note and I'll go through it carefully.

**Comment by Seanny123** [comment]
Yeah, there's quite a few them in the newly created files and they're all
able to be caught by `jscs`, so doing a pass-through would be greatly
appreciated.

File: nengo_gui/static/side_menu.js [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        return item.hidden == false
+    });
+    var last_tab = shown_tabs[shown_tabs.length - 1];
```
Comment (Seanny123) [inline nengo_gui/static/side_menu.js]: I'm uncertain about this logic. Can you have multiple shown tabs? Can you just save the last shown tab before a state change?

**Comment by studywolf** [comment]
Oooooooh this is nice! Fixing that stupid disappearing drop down is awesome.

**Comment by Seanny123** [comment]
Your last commit broke some tests.

Also, can I just make a commit fixing all the styling? Not being able to read the code is making it harder to review.

**Comment by tcstewar** [comment]
> Your last commit broke some tests.

Working on that now....

> Also, can I just make a commit fixing all the styling? Not being able to read the code is making it harder to review.

If you're up for that, that'd be wonderful....

**Comment by studywolf** [comment]
Been playing around with it a bit, it's nice! One thought i had was it
would be really nice to be able to hit escape and exit back out of the menu
instead of having to click that close button!

**Comment by Seanny123** [comment]
Did my best at reformatting the new code. Should be easier to review now.

**Comment by tcstewar** [comment]
> Been playing around with it a bit, it's nice! One thought i had was it would be really nice to be able to hit escape and exit back out of the menu instead of having to click that close button!

Hey @studywolf , could you take a quick look at the current version?  I merged the side-nav tabs with the top toolbar, and I really like the result.  

Right now, closing the menu is done by clicking on the button in the top menu again, but I'll look into adding an escape key too....

File: nengo_gui/static/side_menu.js (hunk 6 above)
Comment (tcstewar) [inline nengo_gui/static/side_menu.js]: Yeah, I'm pretty uncertain about that too....  so I ripped out all the hidden logic stuff.  If you want to hide a tab, just set the class to hidden, rather than mucking about with all this weird indexing stuff.

File: nengo_gui/static/side_menu.js (hunk 4 above)
Comment (tcstewar) [inline nengo_gui/static/side_menu.js]: While getting rid of the hidden weirdness, this turned into a one-liner!  :)

File: nengo_gui/static/side_menu.js [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+Nengo.SideMenu.prototype.hide_side_nav = function() {
+    if (this.menu_open) {
+        this.toggle_side_nav();
+    }
+}
+
+
+//** Determines which tab should be in view when clicked*/
```
Comment (Seanny123) [inline nengo_gui/static/side_menu.js]: Nitpick: this comment is now inaccurate

**Comment by Seanny123** [comment]
This code and interface is beautiful now! Let's fix the failing tests and then open an issue for adding a hotkey to close the menu. After that, I think it's ready to merge.

**Comment by tcstewar** [comment]
> This code and interface is beautiful now! Let's fix the failing tests and then open an issue for adding a hotkey to close the menu. After that, I think it's ready to merge.

Yay!   I just did one more commit to get rid of some extraneous stuff from the old tabs that was still lying around, and that simplifies the code even more.  Yay!  

As for the tests, I'm really confused as to why they're failing, and why it seems to be kinda random.  Those tests pass fine on my local computer, and they don't have anything to do with what's changed here...  hmm....

**Comment by tcstewar** [comment]
The tests pass!  Yay adding more waiting time.  :)

**Comment by Seanny123** [comment]
A corner case that I just thought of: what happens if you try to use the Drag and Drop functionality in the iPython notebook? Can this be disabled in the iPython notebook? Is it already disabled?

**Comment by tcstewar** [comment]
> A corner case that I just thought of: what happens if you try to use the Drag and Drop functionality in the iPython notebook? Can this be disabled in the iPython notebook? Is it already disabled?

That is an excellent question, but I think the right solution to it may interact with the server refactoring, since @jgosmann has cleaned up a lot of the code that determines whether we're running the GUI in different contexts.  So let's hold off on fixing that until both of those are merged.

File: nengo_gui/components/sim_control.py [hunk 8]
```diff
@@ -181,7 +181,8 @@ def javascript(self):
         fn = json.dumps(self.page.filename)
         js = self.javascript_config(info)
         return ('sim = new Nengo.SimControl(control, %s);\n'
-                'toolbar = new Nengo.Toolbar(%s); ' % (js, fn))
+                'toolbar = new Nengo.Toolbar(%s); '
```
Comment (tbekolay) [inline nengo_gui/components/sim_control.py]: Probably want a `\n` at the end of the second line (not that it will matter to JS but if someone is inspecting the JS returned it might help)

File: nengo_gui/static/side_menu.css [hunk 9]
```diff
@@ -0,0 +1,128 @@
+
```
Comment (tbekolay) [inline nengo_gui/static/side_menu.css]: Unnecessary empty line here?

File: nengo_gui/static/side_menu.css [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+.sidenav-container {
+    height: calc(100% - 120px);
```
Comment (tbekolay) [inline nengo_gui/static/side_menu.css]: I'm guessing this is supported in modern Chrome / Firefox / Safari which is what we're targeting right? Might be useful to explicitly state our minimum browser requirements so we can use things like these without worry.

File: nengo_gui/static/side_menu.css [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    transition: 0.5s;
+    border-right: 1px solid #4D4D4D;
+    font-size: 0px;
+    color: #333;
+    font-family: arial;
+    transform: translate(-250px);
+}
+
+.sidenav-container .tab-content{
+  font-size: 15px;
```
Comment (tbekolay) [inline nengo_gui/static/side_menu.css]: Inconsistent tab-width here... we don't really enforce any CSS style though so probably not worth fixing.

File: nengo_gui/static/side_menu.js [hunk 12]
```diff
@@ -0,0 +1,180 @@
+
```
Comment (tbekolay) [inline nengo_gui/static/side_menu.js]: Extra newline here too

**Comment by tbekolay** [comment]
Looked this over; made a few inline comments, mostly nitpicks.

One more important comment is that this PR still includes the start of the drag and drop stuff. We should probably rip that out and move it to a separate branch to be finalized. Right now you can enable it and use it, but the functionality isn't complete (I don't think? For me it added the objects but they don't show in the GUI, so it's not a great user experience). It's also not really the point of this PR, just happened to be developed at the same time.

The side nav itself is a big improvement though, looks quite slick ;)

**Comment by tcstewar** [comment]
> One more important comment is that this PR still includes the start of the drag and drop stuff. We should probably rip that out and move it to a separate branch to be finalized. Right now you can enable it and use it, but the functionality isn't complete (I don't think? For me it added the objects but they don't show in the GUI, so it's not a great user experience). It's also not really the point of this PR, just happened to be developed at the same time.

Very good point.  I was avoiding ripping that out because I didn't have the git knowledge to do it, but after the experience of cleaning up this branch, I felt confident enough to give it a try.  :)  The drag-and-drop is now all in a separate branch (drag-and-drop), and it has been removed from this one.  Yay!

**Comment by tcstewar** [comment]
I'm leaving the styling changes for the PR after this one and the server refactoring get done.  So other than that, I think I'm just waiting to make sure these tests pass, and then I get to have fun merging.  :)

**Comment by tcstewar** [comment]
This has now been merged into master!

However, I screwed up somewhere and ended up with a merge commit.  That didn't happen last time I tried this, and I thought I followed the instructions:

```
(For maintainers) Clean up commits on a branch (prepping for merge)

git checkout <branchtomerge>
git rebase master
git rebase -i <commit>  # if you want to edit the history


(For maintainers) Merge a cleaned up branch

git checkout master
git merge <cleanbranch>
# Ensure that the history looks clean before pushing!
```

@tbekolay any idea what I did wrong?

**Comment by tcstewar** [comment]
Argh, I know what I did wrong....  I didn't do a rebase to master first.  Damnit, sorry about that.  Hmm...  I think I'll wipe out those commits and try it again.....

**Comment by tbekolay** [comment]
Hm... my guess if your master wasn't up to date. Even if you do a `git pull`, master doesn't automatically update, you have to checkout master and `git pull origin master`. I didn't update that doc, but nowadays I do `git merge --ff-only branch` when I'm merging feature branches because of little things like this! Also, setting up git to do rebase pulls by default probably would have solved the issue.

At this point probably no one's pulled so you hard reset back to before the merge and try again.

**Comment by tbekolay** [comment]
Also make sure you force push the rebased version of `side-nav` so that this shows up as merged once you push master.

**Comment by tcstewar** [comment]
Sweet!  I think I fixed it!  :)  thank you @tbekolay !  :)
